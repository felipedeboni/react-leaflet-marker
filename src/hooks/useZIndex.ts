import { useState } from "react";

const useZIndex = (latitudeOffset: number, zIndexOffset: number, riseOffset: number, riseOnHover: boolean) => {
    const [bringToFront, setBringToFront] = useState(false);

    const zIndex: number = latitudeOffset + zIndexOffset + (bringToFront ? riseOffset : 0);

    const events: {
        onMouseEnter: () => void,
        onMouseLeave: () => void,
    } | null = riseOnHover ? {
        onMouseEnter: () => setBringToFront(true),
        onMouseLeave: () => setBringToFront(false),
    } : null;

    return [zIndex, events] as const;
};

export default useZIndex;