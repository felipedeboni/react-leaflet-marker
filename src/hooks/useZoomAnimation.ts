import { DomUtil, LatLngExpression, Point, ZoomAnimEventHandlerFn } from "leaflet";
import { RefObject, useCallback, useEffect } from "react";
import { useMap } from "react-leaflet";

const useZoomAnimation = (
    placementPoint: Point,
    position: LatLngExpression,
    ref: RefObject<HTMLDivElement>
) => {
    const map = useMap();
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const hasAnimation = map._zoomAnimated && map.options.markerZoomAnimation;

    const animateZoom: ZoomAnimEventHandlerFn = useCallback((e) => {
        if (!ref.current) return;

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const point = map._latLngToNewLayerPoint(position, e.zoom, e.center).subtract(placementPoint).round();

        DomUtil.setPosition(
            ref.current,
            point,
        );
    }, [map, placementPoint, position, ref]);

    useEffect(() => {
        if (!hasAnimation) return;
        map.on('zoomanim', animateZoom);

        return () => {
            map.off('zoomanim', animateZoom);
        }
    });

    return hasAnimation;
}

export default useZoomAnimation;