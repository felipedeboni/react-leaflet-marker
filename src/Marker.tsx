import classNames from "classnames";
import React, {FC} from 'react';

import useLatitudeOffset from "./hooks/useLatitudeOffset";
import useMarkerRef from "./hooks/useMarkerRef";
import usePlacementAsPoint from "./hooks/usePlacementAsPoint";
import useUpdatePosition from "./hooks/useUpdatePosition";
import useZIndex from "./hooks/useZIndex";
import useZoomAnimation from "./hooks/useZoomAnimation";
import styles from './Marker.module.css';
import { EPlacement, IMarkerProps } from "./Marker.types";


const Marker: FC<IMarkerProps> = ({
    children,
    position,
    size,
    innerRef,
    riseOffset = 250,
    zIndexOffset = 0,
    interactive= false,
    riseOnHover = false,
    placement = EPlacement.top,
}) => {
    const markerRef = useMarkerRef(innerRef);
    const placementAsPoint = usePlacementAsPoint(placement, size);
    const latitudeOffset = useLatitudeOffset(placementAsPoint, position);
    const hasAnimation = useZoomAnimation(placementAsPoint, position, markerRef);
    const [ zIndex, domEvents ] = useZIndex(
        latitudeOffset,
        zIndexOffset,
        riseOffset,
        riseOnHover
    );
    useUpdatePosition(
        placementAsPoint,
        position,
        markerRef,
    );

    return (
        <div
            ref={markerRef}
            onMouseEnter={domEvents?.onMouseEnter}
            onMouseLeave={domEvents?.onMouseLeave}
            className={classNames(styles.wrapper, {
                [styles.wrapper_interactive]: interactive,
                ['leaflet-zoom-animated']: hasAnimation,
                ['leaflet-zoom-hide']: !hasAnimation,
            })}
            style={{
                zIndex,
                ...(size ? {
                    width: size[0],
                    height: size[0],
                } : null),
            }}
        >
            {children}
        </div>
    );
}

export default Marker;