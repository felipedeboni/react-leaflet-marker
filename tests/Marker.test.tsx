import  "@testing-library/jest-dom";

import { cleanup, fireEvent, render } from "@testing-library/react";
import React from 'react'

import useZIndex from '../src/hooks/useZIndex';
import useZoomAnimation from '../src/hooks/useZoomAnimation';
import Marker from '../src/Marker'

jest.mock('../src/hooks/useMarkerRef', () => jest.fn());
jest.mock('../src/hooks/useLatitudeOffset', () => jest.fn());
jest.mock('../src/hooks/usePlacementAsPoint', () => jest.fn());
jest.mock('../src/hooks/useUpdatePosition', () => jest.fn());
jest.mock('../src/hooks/useZIndex', () => jest.fn(() => [123, null]));
jest.mock('../src/hooks/useZoomAnimation', () => jest.fn(() => false));

describe('Marker', () => {
    beforeEach(() => {
        (useZIndex as jest.Mock).mockReturnValue([123, null]);
        (useZoomAnimation as jest.Mock).mockReturnValue(false);
    });
    afterEach(cleanup);

    test('check zIndex', () => {
        (useZIndex as jest.Mock).mockReturnValue([123, null]);
        const { container } = render(
            <Marker
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                <div>children</div>
            </Marker>
        );

        expect(container.firstChild).toHaveStyle({
            zIndex: 123,
        });
    });

    test('check size styles', () => {
        const { container } = render(
            <Marker
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
                size={[40, 50]}
            >
                <div>children</div>
            </Marker>
        );

        expect(container.firstChild).toHaveStyle({
            width: '40px',
            height: '50px',
        });
    });

    test('have to has classNames', () => {
        (useZoomAnimation as jest.Mock).mockReturnValue(true);
        const { container, rerender } = render(
            <Marker
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                <div>children</div>
            </Marker>
        );
        expect(container.firstChild).toHaveClass('wrapper');
        expect(container.firstChild).not.toHaveClass('wrapper_interactive');

        expect(container.firstChild).toHaveClass('leaflet-zoom-animated');
        expect(container.firstChild).not.toHaveClass('leaflet-zoom-hide');

        (useZoomAnimation as jest.Mock).mockReturnValue(false);
        rerender(
            <Marker
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                <div>children</div>
            </Marker>
        );

        expect(container.firstChild).not.toHaveClass('leaflet-zoom-animated');
        expect(container.firstChild).toHaveClass('leaflet-zoom-hide');

        rerender(
            <Marker
                interactive={true}
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                <div>children</div>
            </Marker>
        );

        expect(container.firstChild).toHaveClass('wrapper_interactive');
    });

    test('zIndex events', () => {
        const onMouseEnter = jest.fn();
        const onMouseLeave = jest.fn();
        (useZIndex as jest.Mock).mockImplementation(() => [231, {
            onMouseEnter,
            onMouseLeave,
        }]);
        const { getByText } = render(
            <Marker
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                just text
            </Marker>
        );
        const content = getByText(/just text/);

        fireEvent.mouseEnter(content);
        fireEvent.mouseLeave(content);
        fireEvent.mouseEnter(content);

        expect(onMouseEnter).toHaveBeenCalledTimes(2);
        expect(onMouseLeave).toHaveBeenCalledTimes(1);
    });

    test('Snapshot', () => {
        const { asFragment } = render(
            <Marker
                interactive
                size={[23,54]}
                position={{lat: 55.84294011297764, lng: 48.52798461914063}}
            >
                children
            </Marker>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});