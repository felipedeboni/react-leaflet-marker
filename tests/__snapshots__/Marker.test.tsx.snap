// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker Snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper wrapper_interactive leaflet-zoom-hide"
    style="z-index: 123; width: 23px; height: 54px;"
  >
    children
  </div>
</DocumentFragment>
`;
